torch/fx/passes/runtime_assert.py

```diff
@@ -347,51 +373,573 @@ def go(node, keypath):
+                            torch.ops.aten.sym_size.int, (node, i)
+                        ),
+                    )
+                for i, s in enumerate(t.stride()):
+                    match_symbol(
+                        s,
+                        lambda: graph.call_function(
+                            torch.ops.aten.sym_stride.int, (node, i)
+                        ),
+                    )
+                match_symbol(
+                    t.storage_offset(),
+                    lambda: graph.call_function(
+                        torch.ops.aten.sym_storage_offset.default, (node,)
+                    ),
+                )
+
+        # if call_function node produces a symbolic value, hash cons
+        if (
+            node.op == "call_function"
+            and (sym_val := _get_sym_val(node)) is not None
+        ):
+            if sym_val in hash_cons:
+                hash_node = hash_cons[sym_val].node
+                node.replace_all_uses_with(hash_node)
+                node.graph.erase_node(node)
+                log.debug("CSEd node %s -> %s for expr %s", node, hash_node, sym_val)
+                continue
+                # I don't think we need to run DCE here...
+                # this would only be needed if the graph has multiple ways
+                # of constructing the same expr for some reason
+                # e.g. x + y + z -> (x + y) + z, or x + (y + z)
+                # -> we would DCE some inputs for the newly-hashed branch
+                # but I don't think this is needed for us
+            elif not _is_size_function_node(node):
+                # we don't want size calls on intermediate tensors, so skip this
+                hash_cons[sym_val] = fx.Proxy(node)
+
+        # if size call on intermediate tensor, convert to compute on input sizes
+        # e.g. y = torch.cat([x, x], dim=1), z = y.shape[0]
+        # convert from: z = sym_size.int(y, 0)
+        # to: x0 = sym_size.int(x, 0), z = x0 * 2
+        # so we can clean up intermediate tensors from memory earlier.
+        if (
+            _is_size_function_node(node)
+            and node.args[0].op != "placeholder"
+            and (sym_shape := _get_sym_val(node)) is not None
+        ):
+            # if this sym expr doesn't exist, or it does exist and it's a size call,
+            # then we create/overwrite by calling sympy_interp()
+            if (
+                sym_shape not in hash_cons
+                or _is_size_function_node(hash_cons[sym_shape].node)
+            ):
+                hash_cons[sym_shape] = sympy_interp(
+                    PythonReferenceAnalysis, hash_cons, sym_shape, hash_cons=hash_cons, insert_after_args=True,
+                )
+            hash_node = hash_cons[sym_shape].node
+            node.replace_all_uses_with(hash_node)
+            log.debug("CSEd intermediate size node for expr %s", sym_shape)
+            _dce_from_node(node)  # here we should DCE
```

Comment:
Why can't we just call DCE all the way in the end?

Reply:
We can, this just doesn't requiring iterating over all the nodes to do so, since we know what to DCE on. But maybe that is cleaner..